### PR TITLE
Remove `InternedString`

### PR DESCRIPTION
**Pull request by nnethercote**
This PR removes `InternedString` by converting all occurrences to `Symbol`. There are a handful of places that need to use the symbol chars instead of the symbol index, e.g. for stable sorting; local conversions `LocalInternedString` is used in those places.

r? @eddyb

### CODE REVIEW

File: src/librustc/hir/map/definitions.rs [hunk 1]
```diff
@@ -136,7 +136,9 @@ impl DefKey {
 
         ::std::mem::discriminant(data).hash(&mut hasher);
         if let Some(name) = data.get_opt_name() {
-            name.hash(&mut hasher);
+            // Get a stable hash by considering the symbol chars rather than
+            // the symbol index.
+            name.as_str().hash(&mut hasher);
```
Comment (eddyb) [inline src/librustc/hir/map/definitions.rs]: Is there no more straight-forward `.stable_hash` or similar?

Comment (nnethercote) [reply src/librustc/hir/map/definitions.rs]: There is `Symbol::hash_stable` but it requires a `StableHashingContext` argument, even though it doesn't use that argument. This code is equivalent.

**Review by eddyb — APPROVED** [review]

**Comment by eddyb** [comment]
Oh, btw, I think `LocalInternedString` would make more sense if it was called `SymbolContents`.

**Comment by nnethercote** [comment]
> Oh, btw, I think `LocalInternedString` would make more sense if it was called `SymbolContents`.

I agree! I was planning to rename in a follow-up. I had thought of `SymbolChars` but `SymbolContents` is also good.

**Comment by nnethercote** [comment]
BTW, you can now do `!bors try !rust-timer queue` (but with `@` instead of `!`) to queue a try run and a perf run in one command.

**Comment by eddyb** [comment]
@rust-timer queue

**Comment by Mark-Simulacrum** [comment]
Queue needs to be run before the try build completes; @rust-timer build c7f73500c3bcd5b6268b7c1ed4b14d78ee82ce59

**Comment by nnethercote** [comment]
Instruction counts look good, lots of small reductions of up to 1%. That's not surprising; hashing and ordering the symbol index is clearly cheaper than the symbol chars (which also involve a TLS lookup).

**Comment by michaelwoerister** [comment]
Just to be sure: `Hash` and `Eq` are still based on the same data for each type? If the two disagree it can lead to subtle, hard to detect bugs.

**Comment by nnethercote** [comment]
> Just to be sure: `Hash` and `Eq` are still based on the same data for each type? If the two disagree it can lead to subtle, hard to detect bugs.

For `Symbol` both `Hash` and `Eq` are derived, and therefore both work with the index. And `LocalInternedString` doesn't implement `Hash`. So everything should be fine.

**Comment by michaelwoerister** [comment]
Great! Thanks for confirming, @nnethercote.

**Comment by nnethercote** [comment]
@eddyb gave me r+ on Discord.

@bors r=eddyb

**Pull request merged**